WIP: Improve documentation (#2017)
ansible-community/ansible-lint: docs/philosophy.md

### What if I do not agree with a specific rule?

We recognize that some projects will find at least one rule that might not suit their needs. Use the `skip_list` feature to temporarily bypass that rule until you have time to update your code.

### Who decides on best practices are adopted?

The main source of new ideas was and remains our community. Before proposing a change check with few other Ansible users that work on different projects and see if they find it useful or not.

It is better to get enough relevant feedback on our discussions forum about before starting to implement new rules. If the proposed rule appears popular and is not conflicting with existing rules, a core will tell you that you start working on it.

The core team will decide on how a new rule will be added. Usually they are added as experimental (warnings only) or even as opt-ins, being made implicit only when a major version is released.

### Do I need to pass all rules to get my collection certified?

Not really. The certification process is likely to use only a subset of rules. At this time, we are working on building that list.

### Why lots of official Ansible docs examples are not passing linting?

> Most of the official examples are written in order to exemplify specific features, and some might conflict with our rules. Still, we plan to include linting of official examples in the future and add specific exclusions where needed, making more likely that a copy/paste from the docs will not raise a bunch of linter violations.

Reviewer: ```suggestion
Most of the official examples are written in order to exemplify specific features, and some might conflict with our rules. Still, we plan to include linting of official examples in the future and add specific exclusions where needed, making it more likely that a copy/paste from the docs will not raise a bunch of linter violations.
```

# Philosophy of ansible-lint

Ansible-lint should be seen as a guru, that helps users write better ansible
content. Still, take all its advice with a grain of salt.

Regarding customizations, we share similar goals with black and prettier, so
we will **avoid adding options** that would allow users to go into different
directions. We hate entropy in our source code, and we will do our best to
help people make code look similar, regardless if it was written by one
author or another.

We do not believe that formatting source code an art, so in order to save you
from wasting time during code reviews about code style or formatting, we will
do our best to do it for you. At least, is better to blame a tool for being
opinionated than antagonizing your peers.

## History and the future

Ansible-lint is almost a decade old and its current list of rules is the result

> of a collaboration of lots of people and not a single person or team. It originated

Reviewer 2: So many "of", how about:

```suggestion
of a collaboration between lots of people and not a single person or team. It originated
```

outside Ansible but we adopted it in 2020 and it was used single by our ansible
galaxy importer.

> For the future, we plan to make it official part of Ansible Automation Platform,

Reviewer: ```suggestion
For the future, we plan to make it an official part of Ansible Automation Platform,
```
or
```suggestion
For the future, we plan to make it officially part of Ansible Automation Platform,
```

and use it in order to certify collections.

## Q&A

> ### Why linter does not accept valid ansible syntax?

Reviewer 3: ```suggestion
### Why does the linter not accept valid ansible syntax?
```

> The linter will always be more restrictive regarding what it accepts. It is
> part of its design. We are not forced to keep the same backwards compatibility

Reviewer: ```suggestion
The linter will always be more restrictive regarding what it accepts. That is
part of its design. We are not forced to keep the same backwards compatibility
```

level as Ansible, so we can tell people to avoid specific syntax for various
reasons, like being deprecated, unsafe or hard to maintain.

The linter will nag you about unrecommended things, even before ansible-core
will start doing so.

> ### What if I do not agree with specific rule?

Reviewer: ```suggestion
### What if I do not agree with a specific rule?
```

> We are confident that most projects with find at least one rule that might not
> suit their needs. We expect you to make use of `skip_list` feature, hopefully
> only temporary until you have time to fix your code.

Reviewer: ```suggestion
We are confident that most projects will find at least one rule that might not
suit their needs. We expect you to make use of the `skip_list` feature, hopefully
only temporarily until you have time to fix your code.
```

### Do I need to pass all rules to get my collection certified?

> Not really. Certification process is very likely to use only a subset of

Reviewer 3: ```suggestion
Not really. Certification process is very likely going to use only a subset of
```

rules and not all of them. Still, at this we are unable to provide the exact
list.

> ### Why lots of official ansible docs examples are not passing linting?

Reviewer: ```suggestion
### Why lots of official ansible docs examples do not pass linting?
```

> ### Why lots of official ansible docs examples are not passing linting?

Reviewer 3: ```suggestion
### Why do lots of official ansible docs examples not pass linting?
```

Most of the official examples are written in order to exemplify specific
features, and some might conflict with our rules. Still, we plan to include
linting of official examples in the future and add specific exclusions where
needed, making more likely that a copy/paste from the docs will not raise a
bunch of linter violations.

> ### Linter requires an ansible version newer than what I use in production?

Reviewer: ```suggestion
### Why does the linter require a newer ansible version than what I use in production?
```

> Linter is used as a static analysis tool and it is perfectly normal to run it
> with a version of ansible that is different than what you would use in
> production.

Reviewer: ```suggestion
The linter is used as a static analysis tool and it is perfectly normal to run
it with a version of ansible that is different than what you would use in
production.
```

> It is part of its goals to help you prepare your code for the future, so don't
> be afraid of using it in such way.

Reviewer: ```suggestion
One of the linter's goals is to help you prepare your code for the future, so don't
be afraid of using it with newer ansible versions.
```
